lib/Analysis/NoiseAnalysis/BGV/NoiseByVarianceCanEmbModel.cpp

```diff
@@ -0,0 +1,300 @@
+  auto std0 = param.getSchemeParam()->getStd0();
+  // probability of larger than 6 * std0 is less than 2^{-30}
+  auto assurance = 6;
+  auto boundErr = std0 * assurance;
+  return boundErr;
+}
+
+template <bool P>
+double Model<P>::getBoundKey(const LocalParamType &param) {
+  // assume UNIFORM_TERNARY
+  auto boundKey = 1.0;
+  return boundKey;
+}
+
+template <bool P>
+double Model<P>::getVarianceErr(const LocalParamType &param) {
+  auto std0 = param.getSchemeParam()->getStd0();
+  return std0 * std0;
+}
+
+template <bool P>
+double Model<P>::getVarianceKey(const LocalParamType &param) {
+  // assume UNIFORM_TERNARY
+  return 2.0 / 3.0;
+}
+
+template <bool P>
+double Model<P>::getFactorD(const LocalParamType &param) {
+  // probability that a exceeds its standard deviation by more than a factor of
+  // D is roughly erfc(D) with erfc(6) = 2^-55, erfc(5) = 2^-40, erfc(4.5) =
+  // 2^-32
+  return 6.;
+}
+
+template <bool P>
+double Model<P>::getBScale(const LocalParamType &param) {
+  auto varianceKey = getVarianceKey(param);
+  auto t = param.getSchemeParam()->getPlaintextModulus();
+  auto d = getFactorD(param);
+  auto phi = getPhi(param);
+
+  // B_scale = D * t * sqrt(phi(m)/12 * (1 + phi(m) * V_key)
+  double innerTerm = (phi / 12.) * (1 + phi * varianceKey);
+  return d * t * sqrt(innerTerm);
+}
+
+template <bool P>
+double Model<P>::getBKs(const LocalParamType &param) {
+  auto varianceError = getVarianceErr(param);
+  auto t = param.getSchemeParam()->getPlaintextModulus();
+  auto d = getFactorD(param);
+  auto phi = getPhi(param);
+
+  // B_ks = D * t * phi(m) * sqrt(V_err / 12)
+  return d * t * phi * sqrt(varianceError / 12.);
+}
+
+template <bool P>
+double Model<P>::getPhi(const LocalParamType &param) {
+  return param.getSchemeParam()->getRingDim() / 2.;
+}
+
+template <bool P>
+typename Model<P>::StateType Model<P>::evalEncryptPk(
+    const LocalParamType &param) {
+  auto varianceError = getVarianceErr(param);
+  // uniform ternary
+  auto varianceKey = getVarianceKey(param);
+  auto t = param.getSchemeParam()->getPlaintextModulus();
+  auto d = getFactorD(param);
+  auto phi = getPhi(param);
+
+  // public key (-as + t * e, a)
+  // public key encryption (-aus + t(u * e + e_0) + m, au + e_1)
+  // ||m + t * (u * e + e_1 * s + e_0)||
+  // <= D * t * sqrt(phi(m) * (1/12 + 2 * phi(m) * V_err * V_key + V_err))
+  double innerTerm =
+      phi * (1. / 12. + 2. * phi * varianceError * varianceKey + varianceKey);
+  double fresh = d * t * sqrt(innerTerm);
+  return StateType::of(fresh);
+}
+
+template <bool P>
+typename Model<P>::StateType Model<P>::evalEncryptSk(
+    const LocalParamType &param) {
+  auto varianceError = getVarianceErr(param);
+  auto t = param.getSchemeParam()->getPlaintextModulus();
+  auto d = getFactorD(param);
+  auto phi = getPhi(param);
+
+  // secret key s
+  // secret key encryption (-as + m + t * e, a)
+  // ||m + t * (u * e + e_1 * s + e_0)||
+  // <= D * t * sqrt(phi(m) * (1/12 + V_err))
```

Comment:
The comment here should be `||m + t * e|| <= D * t * sqrt(phi(m) * (1/12 + Verr))`

Reply:
Changed the comment